hasktorch/src/Torch/Typed/Optim/CppOptim.hs

```diff
@@ -0,0 +1,141 @@
+    => option -> model -> IO (CppOptimizerState option (Parameters model))
+
+  unsafeStep :: forall model dev dtype lossShape tensors res. 
+    (Parameterized model
+    ,HMap' ToDependent (Parameters model) tensors
+    ,HMap' ToParameter tensors (Parameters model)
+    ,Castable (HList tensors) [TD.ATenTensor]
+    )
+      => model 
+      -> CppOptimizerState option (Parameters model)
+      -> Tensor dev dtype lossShape
+      -> IO (model, CppOptimizerState option (Parameters model))
+  unsafeStep model o@(CppOptimizerState _ optimizer) loss = do
+    let deps :: HList tensors
+        deps = hmap' ToDependent $ flattenParameters model
+
+    -- Debug.traceIO $ "Tensors in: "
+    -- cast deps (Debug.traceIO . show . map (TD.shape . TD.Unsafe))
+    v :: [TD.ATenTensor] <- cast3 LibTorch.unsafeStep optimizer deps loss
+    -- Debug.traceIO $ "Params returned by unsafeStep: "<>show (length v)
+    
+    newParamTensors :: HList tensors <- uncast v pure
+    -- Debug.traceIO $ "Tensors out: "
+    -- cast newParamTensors (Debug.traceIO . show . map (TD.shape . TD.Unsafe))
+    let newParams = hmap' ToParameter newParamTensors
+    let newModel = replaceParameters model newParams
+    return (newModel, o)
+
+instance CppOptimizer AdamOptions where
+  initOptimizer  opt@AdamOptions{..} model = do
+    v <- cast7 LibTorch.adam adamLr (fst adamBetas) (snd adamBetas)
+                             adamEps adamWeightDecay adamAmsgrad initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+
+instance CppOptimizer AdamwOptions where
+  initOptimizer  opt@AdamwOptions{..} model = do
+    v <- cast7 LibTorch.adamw adamwLr (fst adamwBetas) (snd adamwBetas) adamwEps adamwWeightDecay adamwAmsgrad initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+
+instance CppOptimizer LbfgsOptions where
+  initOptimizer opt@LbfgsOptions{..} model = do
+    v <- cast8 LibTorch.lbfgs lbfgsLr lbfgsMaxIter lbfgsMaxEval lbfgsToleranceGrad lbfgsToleranceChange lbfgsHistorySize lbfgsLineSearchFn initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+instance CppOptimizer RmspropOptions where
+  initOptimizer opt@RmspropOptions{..} model = do
+    v <- cast7 LibTorch.rmsprop rmspropLr rmspropAlpha rmspropEps rmspropWeightDecay rmspropMomentum rmspropCentered initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+instance CppOptimizer SGDOptions where
+  initOptimizer  opt@SGDOptions{..} model = do
+    v <- cast6 LibTorch.sgd sgdLr sgdMomentum sgdDampening sgdWeightDecay sgdNesterov initParams'
+    return $ CppOptimizerState opt v
+    where
+      initParams'= hmap' ToDependent $ flattenParameters model
+
+runStep :: (CppOptimizer option, Parameterized model
```

Comment:
This typed runStep is not the same as Torch.Typed.Optim's one.
Can we use type class for runStep?
If there are any problems on design, there is no need to use it.

Reply:
If it doesn't seem to be easy to fix, I think it's okay to merge this pr.

Reply:
The only difference should be the learningRate parameter which I excluded for the reasons discussed in #467.  The learning rate is part of the `options` field of the `optim` argument already.